[Analyzer] Stop weakening the AST

sodic opened the pull request.

Martinsos commented:
We are back at this!
We did quick analysis, seems we have a lot done, we removed the `weaken` function, and we refactored unification to account for that, we checked that evaluator is ok with all of it, and the only thing left is to update tests!
Once we do that we can merge this PR and get back to working on union types.

sodic commented:
We added a test for `TypeChecker/Internal#check`, fixed one test for `TypeChecker/Internal#checkStmnt` and one for `TypeChecker#typeCheck`.

Next time, we should:
- Write the tests for `TypeChecker/Internal#checkIsSubtypeOf`
- Make sure we're happy with test coverage of the rest of the functions in `TypeChecker/Internal`

Martinsos commented:
We wrote tests for `checkIsSubTypeOf` function. Next, we should make sure we are happy with test coverage for the rest of the functions in TypeChecker/Internal, and then we are done with this issue!
